File: src/leetcode/LeetCode16.java
package leetcode;

import java.util.Arrays;

public class LeetCode16 {

    public int threeSumClosest(int[] nums, int target) {

        Arrays.sort(nums);
        int result = nums[0] + nums[1] + nums[2];
        int sum;
        for (int i = 0; i < nums.length - 2; i++) {
            int start = i + 1;
            int end = nums.length - 1;
            while (start < end) {
                sum = nums[i] + nums[start] + nums[end];
                if (sum > target) {
                    --end;
                } else {
                    ++start;
                }
                if (Math.abs(target - sum) < Math.abs(target - result)) {
                    result = sum;
                }
            }
        }
        return result;
    }
}

File: src/leetcode/LeetCode16.go
package main

import "sort"

func threeSumClosest(nums []int, target int) int {
	length := len(nums)
	sort.Ints(nums)
	res := nums[0] + nums[1] + nums[2]
	for i := 0; i < length-2; i++ {
		left := i + 1
		right := length - 1
		for left < right {
			num := nums[i] + nums[left] + nums[right]
			if num > target {
				right--
			} else if num < target {
				left++
			} else {
				return target
			}
			absNum := target - num
			absRes := target - res
			if absNum < 0 {
				absNum = -absNum
			}
			if absRes < 0 {
				absRes = -absRes
			}
			if absNum < absRes {
				res = num
			}
		}
	}
	return res
}

func main() {

}
